Schema nullable refactoring

joaopapereira opened the pull request:
In this PR some refactoring was done to try to address comments on the commit original commit for the nullable schema annotation implementation.

https://github.com/vmware-tanzu/carvel-ytt/commit/66c26bad47cc0d30f4224c205f054a145793d6b0#r44427722
https://github.com/vmware-tanzu/carvel-ytt/commit/66c26bad47cc0d30f4224c205f054a145793d6b0#r44427606

In broad strokes, the annotation concept is now part of the schema types and instead of relying on magic strings we now use `structmeta.AnnotationName` and implemented a method that checks if a type can be nullable.

joaopapereira committed:
Refactor nullable schema annotation

* Add Schema package which depends on template and yamlmeta packages
  The idea here is to move all the type checking to the schema package
  instead of being in yamlmeta and schema.
* Implement a more generic approach to schema annotations based on
   structmeta.AnnotationName.
* Remove magic nullable annotation string in favour of the
  AnnotationSchemaNullable constant though, this practically
  duplicates the IsNullable function
* Add an empty space to the nullable test to ensure whitespace
  independence
* In this commit we also move the Schema interface to the package that
  will need to use it.
* Minor changes were made to the tests to better represent the error
  messages that we expect to have

Co-Authored-By: Dennis Leon <leonde@vmware.com>
Co-authored-By: Steven Locke <slocke@pivotal.io>
Co-Authored-By: Garrett Cheadle <gcheadle@vmware.com>
Co-Authored-By: John Ryan <jtigger@infosysengr.com>

cppforlife commented on pkg/yamlmeta/type.go:
```diff
@@ -45,22 +46,26 @@ type DocumentType struct {
 type MapType struct {
-	Items []*MapItemType
+	Items       []*MapItemType
+	annotations template.NodeAnnotations
```
im -1 on this. this violates layering that's been setup. yamlmeta package should NOT know anything about templating.

joaopapereira replied:
Can you explain what do you mean by layering?

cppforlife replied:
we would like to maintain boundaries between various packages so that there is a clean layering going on in the "cake" (project). bottom layers of the cake should not depend on higher layers of the cake.

one way to "test this" is to imagine if to use particular package in isolation. eg my other project needs to parse yamls and extract metadata (comments). it doesnt need any concept of templating and i want to interpret comments in my own way (eg `# foo` is something special, everything else is plain comment).

the way that ytt cake is set up is something like this:

- yamlmeta - standalone package that knows how to parse YAML and produce YAML AST.
  - no dependency
- template - standalone generic starlark templating with a high level instructions interface for building structures
  - no dependency
- texttemplate/yamltemplate - language specific (YAML, text, may be JSON in future, etc)
  - yamltemplate -> depends on -> yamlmeta, and template
  - texttemplate -> depends on -> template

(more pkgs descriptions here: https://github.com/vmware-tanzu/carvel-ytt/blob/develop/docs/dev.md#source-code-structure).

cppforlife commented:
> Stop yamlmeta package depending on the template package

yamlmeta pkg still continues to depend on template pkg (https://github.com/vmware-tanzu/carvel-ytt/pull/244/commits/6a351599276b8420823d021e37298452a8668bbc#diff-ae30da157e6a7f9c338ef9aeccf6e4f5ae90e3a460619a3a6d87987597ba4153R9)

StevenLocke commented:
Hmmm, that's a good point. That commit was inaccurately named.

More accurately, the template dependency is now only in the schema type construction, it was removed from the more general type.go. Perhaps the schema code doesn't belong in the yamlmeta package to keep this layering, but the "types" needs to be able to check those annotations at some point. It has to be able to read "#@schema/nullable" somewhere.

joaopapereira commented on pkg/cmd/template/cmd.go:
```diff
@@ -5,6 +5,7 @@ package template
 
 import (
 	"fmt"
+	"github.com/k14s/ytt/pkg/schema"
```
Please move the import to the other group

joaopapereira commented on pkg/workspace/data_values_pre_processing.go:
```diff
@@ -5,6 +5,7 @@ package workspace
 
 import (
 	"fmt"
+	"github.com/k14s/ytt/pkg/schema"
```
Please move the import to the other group

joaopapereira commented on pkg/workspace/library_loader.go:
```diff
@@ -5,6 +5,7 @@ package workspace
 
 import (
 	"fmt"
+	"github.com/k14s/ytt/pkg/schema"
```
Please move the import to the other group

joaopapereira commented on pkg/workspace/library_module.go:
```diff
@@ -5,6 +5,7 @@ package workspace
 
 import (
 	"fmt"
+	"github.com/k14s/ytt/pkg/schema"
```
Please move the import to the other group

joaopapereira commented on pkg/workspace/template_loader.go:
```diff
@@ -5,6 +5,7 @@ package workspace
 
 import (
 	"fmt"
+	"github.com/k14s/ytt/pkg/schema"
```
Please move the import to the other group

joaopapereira commented on pkg/yamlmeta/type.go:
```diff
@@ -6,6 +6,11 @@ package yamlmeta
+)
+
+const (
+	AnnotationSchemaNullable structmeta.AnnotationName = "schema/nullable"
```
Feels like this is a schema annotation. Shouldn't this be in the schema package?

joaopapereira reviewed (commented):
I added some comments, specially around import ordering.

Since we created a schema package now I am thinking that maybe we should not have the Types inside yamlmeta, since the Types make sense associated with a schema, not by themselves.

Not sure what implications this would have for yamlmeta and the dependencies between packages.

joaopapereira committed:
Move logic from CheckAllows to Map's CheckType

- Replace CheckAllows function with CheckType for Maps
  This was done because the function was specific to MapType but it was
  part of the generic Type interface
- Change the behavior of key checking on Maps
  A given Map's keys must all be valid before any values are checked
- Add special treatment for Scalars while checking type of Maps and
  Arrays. This allows the error message to be more helpful.

Signed-off-by: John Ryan <jtigger@infosysengr.com>
